### DISCUSSION TITLE
Optimizing nrepl-dict internals

### DISCUSSION THREAD

**Discussion by yuhan0** [clojure-emacs/cider#3711] [Ideas]
`nrepl-dict-get` is probably the #1 called function in all of Cider, so I was surprised to find that its implementation has an inefficient call to `nrepl-dict-contains`, ~~that turns from what should be a O(n) lookup into O(n²).~~ (edit: see correction below) 

https://github.com/clojure-emacs/cider/blob/05e7570e33d1aa795b1a0f85a6aa2162b42b93cd/nrepl-dict.el#L69

Note that this check has no effect when the optional `default` argument is nil, and really only serves to distinguish the case where the key's value is `nil`:

```lisp
(let ((res '(dict "foo" nil)))
  (nrepl-dict-get res "foo" :not-found) ;; => nil
  (nrepl-dict-get res "bar" :not-found) ;; => :not-found

  ;; Note the subtle difference:
  (or (nrepl-dict-get res "foo") :not-found) ;; => :not-found
  (or (nrepl-dict-get res "bar") :not-found) ;; => :not-found
  )
 ```

One simple improvement would be to only perform the contains? check when `default` arg is non-nil, turning it back into O(n) in the 2-argument case.
... In fact as far as I can tell, the 3-argument call to `nrepl-dict-get` is not used at all in the entire Cider and clj-refactor codebases! I guess it was originally written to mirror the semantics of `clojure.core/get`, overlooking the fact that plist lookups don't have the same O(1) performance guarantees as in Clojure.

So another possible change might be to remove the optional argument entirely, or place a deprecation warning for future removal around its use. I believe nREPL responses can't even contain nils in the first place(?), so its usage would basically be equivalent to the regular Elisp idiom of `(or (nrepl-dict-get ..) <default>)`.

**Comment by bbatsov** [comment]
>  ... In fact as far as I can tell, the 3-argument call to nrepl-dict-get is not used at all in the entire Cider and clj-refactor codebases! I guess it was originally written to mirror the semantics of clojure.core/get, overlooking the fact that plist lookups don't have the same O(1) performance guarantees as in Clojure.

Yeah, you're probably right. 

> I believe nREPL responses can't even contain nils in the first place(?)

Indeed. Bencode has no representation for `nil`. It seems to me that this can just be removed, as I just can't see any uses for it. (and you not finding any kind of proves the point) @alexander-yakushev is our master of performance, so you he might have some other ideas about tweaking the performance of `nrepl-dict`.

**Comment by vemv** [comment]
Optimizing things SGTM - thanks!

The `nrepl-dict-get` name always felt awkwardly verbose to me, so if you found some nice naming, we could use that one and leave the old implementation untouched (which gives us some peace of mind)

**Comment by yuhan0** [comment]
I'm cautious both about introducing more API functions needlessly and about making breaking changes to such a internal part of the codebase.. probably the safe thing to do now is raise a deprecation warning on a 3-argument call, and then remove it in the future (Cider 2.0?)

In the meantime I'll remove the argument entirely on my local fork and make sure nothing goes wrong (so far all seems fine)

**Reply by alexander-yakushev** [reply]
No meaningful input from me. Deprecating 3-arity and improving the other one sounds perfect. I agree that adding another low-level function at this point is unnecessary.

**Reply by bbatsov** [reply]
Okay, I guess we have consensus then.

**Comment by yuhan0** [comment]
Oops, I was wrong about the complexity analysis - `contains?` is O(n) after all, not O(n^2) in the size of the dict. 
But it does affect the worst vs average case performance, since it will always have to traverse the entire plist instead of being able to return early if the key is near the start of the list. 

A quick benchmark shows an average speedup factor of ~0.7 (30% decrease) for dicts with 10 keys, and ~0.4 (60% decrease) for dicts with 100 keys.

In the worst-case comparison (key is at the start of dict, N=100), we get a 92% improvement, but I guess most response dicts are relatively small in practice.

**Reply by yuhan0** [reply]
I take back that last bit...  after using the following advice to trace calls to `nrepl-dict-get`, it turns out it's being called dozens of times per command on huge dictionaries with 800+ elements 😧

```lisp
(defvar !ndg-trace nil)

(advice-add 'nrepl-dict-get :around
            (defun qyh/spy-on-ndg (f d k &optional df)
              (if df
                  (error "3-arg call!")
                (push (length d) !ndg-trace)
                (funcall f d k df))))
;; (advice-remove 'nrepl-dict-get #'qyh/spy-on-ndg)

(add-hook 'post-command-hook
          (defun qyh/log-traces ()
            (when !ndg-trace
              (message "called %d times, max N=%d, avg %f"
                        (length !ndg-trace)
                        (apply #'max 0 !ndg-trace)
                        (/ (apply #'+ !ndg-trace)
                           (length !ndg-trace)))
              (setq !ndg-trace nil))))
;; (remove-hook 'post-command-hook #'qyh/log-traces)
```

Here's some sample output (open a clj file, navigating around buffer triggering eldoc, type a few lines inside a defn form, eval)
```
called 602 times, max N=805, avg 16.000000
called 30 times, max N=221, avg 20.000000
called 29 times, max N=221, avg 21.000000
called 1 times, max N=221, avg 221.000000 [3 times]
called 29 times, max N=221, avg 21.000000
called 1 times, max N=221, avg 221.000000 [3 times]
called 29 times, max N=221, avg 21.000000 [3 times]
called 1 times, max N=221, avg 221.000000
called 29 times, max N=221, avg 21.000000
called 50 times, max N=221, avg 33.000000
called 31 times, max N=221, avg 20.000000
called 16 times, max N=805, avg 288.000000
called 1 times, max N=221, avg 221.000000
called 21 times, max N=805, avg 323.000000
called 1 times, max N=221, avg 221.000000 [3 times]
called 21 times, max N=805, avg 323.000000
called 1 times, max N=221, avg 221.000000
called 21 times, max N=805, avg 323.000000
called 1 times, max N=221, avg 221.000000 [5 times]
=> #'repro/foo
called 45 times, max N=11, avg 9.000000
called 3 times, max N=221, avg 221.000000
```

A bit of further digging reveals two main culprits:
- `nrepl-ops` (N=221) 
  - queried by the very frequently called `nrepl-op-supported-p`
  
- `cider-ns-cache` (N=805)
  -  queried by `cider-resolve--get-in` <- `cider-resolve-var` <- `cider--get-symbol-indent`
    - called multiple times (once per sub-expression being indented) by `clojure-indent-line` 

And It turns out this causes serious perceptible lag in some cases, which I had actually encountered before! (only just made the connection):  https://github.com/clojure-emacs/cider/issues/3343#issuecomment-1552973878

Getting a constant-factor speedup here doesn't help much - even after applying the change to `nrepl-dict-get` it turns out that the following combination causes unacceptable amounts of lag.
- `clojure-use-backtracking-indent = t`
- `aggressive-indent-mode`
- Editing a moderately large (50-100 LOC) expression (resulting in that many calls to clojure-indent-line per keystroke)

**Reply by bbatsov** [reply]
Yeah, sadly the indent function has always been too slow to be used with `aggressive-indent-mode` in bigger files.

**Reply by alexander-yakushev** [reply]
Is it possible to use hashtables for big nrepl responses? Or even for all responses?

**Reply by yuhan0** [reply]
I ran the CPU profiler on this for a bit, it turns out that calls to `nrepl-dict-get` alone contribute to ~40% of the lag experienced with aggressive-indent, and probably trigger frequent GC pauses on top of that (~30%). 

It's actually *even worse* than a linear N=800+ traversal, since `cider-var-resolve` does a get-in lookup into `cider-repl-ns-cache` and these nested dicts can have hundreds of elements each! 


```lisp
(with-current-buffer (seq-find (lambda (b)
                                 (with-current-buffer b
                                   (derived-mode-p 'cider-repl-mode)))
                               (buffer-list))
  (nrepl-dict-map (lambda (k v)
                    (length (nrepl-dict-get v "interns")))
                  cider-repl-ns-cache))
;; =>
;; (245 77 23 17 7 55 21 17 31 13 9 13 25 53 217 45 39 71 9 23 81 59 1 195 59 33 7 5 15 21 57 75 45 507 35 27 ...)

```


Definitely a case for hash-tables here, maybe even interning the string keys for faster comparisons with `eq`.

**Reply by yuhan0** [reply]
> Is it possible to use hashtables for big nrepl responses? Or even for all responses?

I don't think the overhead from initializing hash-tables for every single response would be worth it - my experiments above with tracing the calls to `nrepl-dict-get` showed that the practically all other inputs besides those two outliers are response maps with a small number of keys (2 ~ 10). 

Clojure uses N>8 as a transition from array-map to hash-map, I'm guessing the O(n) vs. O(1) tradeoff would be in a similar range for Elisp, probably higher with things being so list-centric.

**Reply by alexander-yakushev** [reply]
I'm pretty sure that hashtables will get you 90% there, interning can be considered separately.

I suspect that changing all nrepl responses from plists to hashtables might be too heavy-handed (@bbatsov please tell me if I'm wrong). In that case, hashtablizing just those large caches might work.

**Reply by yuhan0** [reply]
Ok, continuing the discussion from #3713, I think there are 2 approaches we could take here - let me know which would be preferable:


### Overloading all nrepl-dict functions to work with hash-tables, in addition to plist-based impl.

This would be a additive (non-breaking) change to the nrepl-dict API, and allow us to switch transparently to a more efficient implementation for dicts above a certain size.

```lisp

(defun nrepl-dict (&rest key-vals)
  (if (<= 100 (length key-vals))
      <<make-hash-table>>
    (cons 'dict key-vals)))

(defun nrepl-dict-get (dict key)
  (cond ((null dict) nil)
        ((nrepl-dict-p dict)
         <<current-implementation>>)
        ((hash-table-p dict)
         (gethash key dict))
        (t
         (error "Not an nREPL dict object: %s" dict))))

;; eg. both the following would work
(nrepl-dict-get '(dict ...) k) => v
(nrepl-dict-get #s(hash-table ...) k) => v

```

But there's the possibility that this would introduce subtle bugs via some code which directly uses car/cdr/plist-methods on a dict, and in someone's specific setup, it happens to cross the threshold and gets passed a hash-table instead.


### Targeted refactoring of `nrepl-ops` and `cider-repl-ns-cache`

These appear to be the only major outliers with N>100 elements, so we could leave the nrepl-dict API unchanged for use with small response maps, and specifically refactor these two variables and their surrounding code to use hash-tables.

This has the opposite risk of breaking external packages / user snippets which directly interact with those variables through the regular nrepl-dict interfaces...

A Github code search turned up nothing, so it's probably safe? Being buffer-local variables in the REPL connection buffer, one can consider them as private-ish internal vars, even if they aren't named with the `--` convention.

**Reply by alexander-yakushev** [reply]
I actually like option number 1 better. Since CIDER always used these specialized `nrepl-dict-*` accessors, it was kinda implied to treat the data values underneath as a black box.

**Reply by bbatsov** [reply]
> A Github code search turned up nothing, so it's probably safe? Being buffer-local variables in the REPL connection buffer, one can consider them as private-ish internal vars, even if they aren't named with the -- convention.

Indeed, those are not meant to be used directly and I'm reasonably sure no one touches them directly. So probably that'd be the easier approach and my vote goes to it.

**Reply by bbatsov** [reply]
> I actually like option number 1 better. Since CIDER always used these specialized `nrepl-dict-*` accessors, it was kinda implied to treat the data values underneath as a black box.

Yeah, that's a fair point. In general I don't have any objections for both approaches, it just seems to me that the second will be a bit less work. I'm fine with either.

**Reply by yuhan0** [reply]
Ah, here's one such case which might break under option 1:
https://github.com/clojure-emacs/cider/blob/105da319b09a436552f1b3c6194cbbc833017dd2/nrepl-client.el#L1320

(found by grepping for `lax-plist-get`, there may be other less obvious instances of this sort of thing)

**Reply by yuhan0** [reply]
I'll go with option 2 then, probably dogfooding it for a week or two before submitting a PR.
